iree/compiler/Dialect/HAL/Conversion2/StreamToHAL/ConvertStreamToHAL.cpp

```diff
@@ -0,0 +1,1149 @@
+
+    // TODO(benvanik): propagate semaphore information.
+    auto resolvedTimepoint =
+        rewriter.create<arith::ConstantIndexOp>(loc, 0).getResult();
+
+    rewriter.replaceOp(executeOp, resolvedTimepoint);
+    return success();
+  }
+};
+
+struct CmdSerialOpPattern
+    : public StreamConversionPattern<IREE::Stream::CmdSerialOp> {
+  using StreamConversionPattern::StreamConversionPattern;
+  LogicalResult matchAndRewrite(
+      IREE::Stream::CmdSerialOp serialOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto commandBuffer = mapping->lookupCommandBufferFor(serialOp);
+
+    // Run through the execution region and serialize execution by inserting
+    // barriers. Nested regions may elide barriers as needed.
+    auto &bodyBlock = serialOp.body().front();
+    insertSerializationBarriers(serialOp.getLoc(), bodyBlock, commandBuffer,
+                                OpBuilder::atBlockBegin(&bodyBlock));
+
+    // Inline the serial execution region.
+    rewriter.mergeBlockBefore(&serialOp.body().front(), serialOp);
+    rewriter.eraseOp(serialOp);
+    return success();
+  }
+};
+
+struct CmdConcurrentOpPattern
+    : public StreamConversionPattern<IREE::Stream::CmdConcurrentOp> {
+  using StreamConversionPattern::StreamConversionPattern;
+  LogicalResult matchAndRewrite(
+      IREE::Stream::CmdConcurrentOp concurrentOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    // Inline the concurrent execution region.
+    // TODO(benvanik): split barriers (event set/wait) when nesting.
+    rewriter.mergeBlockBefore(&concurrentOp.body().front(), concurrentOp);
+    rewriter.eraseOp(concurrentOp);
+    return success();
+  }
+};
+
+struct TimepointImmediateOpPattern
+    : public StreamConversionPattern<IREE::Stream::TimepointImmediateOp> {
+  using StreamConversionPattern::StreamConversionPattern;
+  LogicalResult matchAndRewrite(
+      IREE::Stream::TimepointImmediateOp immediateOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    // TODO(benvanik): model timepoints as semaphores.
+    rewriter.replaceOpWithNewOp<arith::ConstantIndexOp>(immediateOp, 0);
+    return success();
+  }
+};
+
+struct TimepointJoinOpPattern
+    : public StreamConversionPattern<IREE::Stream::TimepointJoinOp> {
+  using StreamConversionPattern::StreamConversionPattern;
+  LogicalResult matchAndRewrite(
+      IREE::Stream::TimepointJoinOp joinOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    // TODO(benvanik): model timepoints as semaphores.
+    // This should be a max() of the operand timepoints. MLIR has no max op
+    // until https://reviews.llvm.org/D110540. Could be done with affine
+    // expressions, but since everything is always 0 we just max(0,0)=0 here :)
+    rewriter.replaceOpWithNewOp<arith::ConstantIndexOp>(joinOp, 0);
```

Comment:
The linked https://reviews.llvm.org/D110540 has landed, FYI. (fine to address in a follow-up)